packages/libraries/cli/src/helpers/output-schema.ts

```diff
@@ -0,0 +1,41 @@
+import { z } from 'zod';
+
+// prettier-ignore
+export type OutputSchema =
+		| Envelope
+		| z.ZodUnion<[Envelope]>
+		| z.ZodUnion<[Envelope, Envelope]>
+		| z.ZodUnion<[Envelope, Envelope, Envelope]>
+		| z.ZodUnion<[Envelope, Envelope, Envelope, Envelope]>;
+// ... as many as needed
+
+type Envelope = typeof OutputSchema.Envelope;
+
+export namespace OutputSchema {
+  export const NonEmptyString = z.string().min(1);
+
+  export const Envelope = z.object({
+    ok: z.literal(true),
+    message: z.string().optional(),
+    // warnings: z.array(z.string()),
+  });
+
+  export namespace Effect {
+    export const Skipped = Envelope.extend({
+      effect: z.literal('skipped'),
+    });
+    export const Executed = Envelope.extend({
+      effect: z.literal('executed'),
+    });
+  }
+  export namespace DataOutputMode {
+    export const Stdout = z.object({
+      outputMode: z.literal('stdout'),
+      content: z.string(),
+    });
+    export const File = z.object({
+      outputMode: z.literal('file'),
+      path: z.string(),
+    });
+  }
+}
```

Comment:
Can you give some context on using namespaces here instead of just exporting an object?

```ts
export const Effect = {
  Skipped: Enevlop.extend({ /* ... */ }),
  /* ... */
  DataOutputMode: {
     Stdout: z.object({ /* ... */ }),
     /* ... */
  },
}
```

This is a new pattern introduced to the code base here, and I wonder about its benefits.

Reply:
The answer to this is subtle.

Conceptually `namespace` model inline modules but in practice they only do so for types whilst for values they actually just compile to objects at runtime (but only _if there are any values exported from them_).

They have a special property in TS which is that they can support exporting types as well. And those type names can overlap with the value names. This _sometimes has advantages for library design_. In practice, what `namespace` can do so can additional ES modules. The "problem" with ES modules is how heavyweight it is to create new files be comparison of inline, single-line modules. And, when ES modules _are_ used, then, you actually deal with a statically analyzable structure, which plain objects are not. This is relevant for Tree shaking for example.

Related to all this is a tc39 proposal for inline modules that, if make it into the language, I suspect will allow the use cases for `namespace` to be entirely replaced by something native to JS, which would be nice https://github.com/tc39/proposal-module-declarations.

In terms of code style, returning to this code block... here is the summary of reasons using `export const Effect ...` is not 1:1:

1. It is true that `namespace` _when there are values exports_ compiles to `export const ...` however that can be seen as an implementation detail. From a readability point of view we can view `namespace` to mean this: "I want this to be a separate namespace but I was too lazy or unsure yet about the direction to warrant creating more files with `* as X` imports etc."

2. As soon as we want to do something as follows then `namespace` (or ESM modules) is the _only way_:

	```ts
	export namespace Effect {
	  export const Skipped = ...
	  export type Skipped = z.infer<typeof Skipped>
	}
    // ...
    // Somewhere else in the codebase:
    // ...
    const foo: Effect.Skipped = ...
	```

In terms of usage, some additional points:

  - There are some cases where IDEs currently have poor intellisense for autocomplete of values/types from ESM namespaces whereas they fair better with `namespace`. This is a consideration.
  - Every single instance of `namespace` is an instance where ESM modules could have been used instead. But `namespace` is a lot easier to change since there's no extra files involved. So sometimes when I'm introducing some new code/ideas, I start with `namespace` because I'm not 100% sure yet how that code will evolve. Mature code in libraries should probably usually go towards ESM modules where tree shaking support is important. For apps, tree-shaking may not matter, in which case ESM modules vs namespaces don't matter much either.